crates/ty_python_semantic/resources/mdtest/function/return_type.md

```diff
@@ -182,6 +182,225 @@ def f(cond: bool) -> int:
+reveal_type(odd(1))  # revealed: bool | Unknown
+
+def repeat_a(n: int):
+    if n <= 0:
+        return ""
+    else:
+        return repeat_a(n - 1) + "a"
+
+# TODO: it may be better to infer this as `str` if we can
+reveal_type(repeat_a(3))  # revealed: Literal[""] | Unknown
+
+def divergent(value):
+    if type(value) is tuple:
+        return (divergent(value[0]),)
+    else:
+        return None
+
+# tuple[tuple[tuple[...] | None] | None] | None => tuple[Unknown] | None
+reveal_type(divergent((1,)))  # revealed: tuple[Unknown] | None
+
+def nested_scope():
+    def inner():
+        return nested_scope()
+    return inner()
+
+reveal_type(nested_scope())  # revealed: Unknown
+
+def eager_nested_scope():
+    class A:
+        x = eager_nested_scope()
+
+    return A.x
+
+reveal_type(eager_nested_scope())  # revealed: Unknown
+```
+
+### Class method
+
+If a method's return type is not annotated, it is also inferred, but the inferred type is a union of
+all possible return types and `Unknown`. This is because a method of a class may be overridden by
+its subtypes. For example, if the return type of a method is inferred to be `int`, the type the
+coder really intended might be `int | None`, in which case it would be impossible for the overridden
+method to return `None`.
+
+```py
+class C:
+    def f(self):
+        return 1
+
+class D(C):
+    def f(self):
+        return None
+
+reveal_type(C().f())  # revealed: Literal[1] | Unknown
+reveal_type(D().f())  # revealed: None | Literal[1] | Unknown
+```
+
+However, in the following cases, `Unknown` is not included in the inferred return type because there
+is no ambiguity in the subclass.
+
+- The class or the method is marked as `final`.
+
+```py
+from typing import final
+
+@final
+class C:
+    def f(self):
+        return 1
+
+class D:
+    @final
+    def f(self):
+        return "a"
+
+reveal_type(C().f())  # revealed: Literal[1]
+reveal_type(D().f())  # revealed: Literal["a"]
+```
+
+- The method overrides the methods of the base classes, and the return types of the base class
+    methods are known (In this case, the return type of the method is the intersection of the return
+    types of the methods in the base classes).
+
+```toml
+[environment]
+python-version = "3.12"
+```
+
+```py
+from typing import Literal
+
+class C:
+    def f(self) -> int:
+        return 1
+
+    def g[T](self, x: T) -> T:
+        return x
+
+class D(C):
+    def f(self):
+        return 2
+    # TODO: This should be an invalid-override error.
+    # If the override is invalid, the type of the method should be that of the base class method.
+    def g(self, x: int):
+        return 2
+
+class E(D):
+    def f(self):
+        return 3
+
+reveal_type(C().f())  # revealed: int
+reveal_type(D().f())  # revealed: int
+reveal_type(E().f())  # revealed: int
+reveal_type(C().g(1))  # revealed: Literal[1]
+# TODO: should be `Literal[1]`
+reveal_type(D().g(1))  # revealed: Literal[2] | T
```

Comment:
I think what we reveal here is not super important, since it's the result of an invalid override, but it is important that we not reveal an unspecialized `T` here. If anything we could just fall back to `Unknown`.